File: src/services/storage.ts
import fs from "fs-extra";
import os from "os";
import path from "path";
import { DobbyState } from "../types/plan";

const DATA_DIR = path.join(os.homedir(), ".dobby-cli");
const DATA_FILE = path.join(DATA_DIR, "state.json");

const defaultState: DobbyState = {
  plan: null,
};

export class StorageService {
  async ensureDataFile(): Promise<void> {
    await fs.ensureDir(DATA_DIR);
    if (!(await fs.pathExists(DATA_FILE))) {
      await fs.writeJson(DATA_FILE, defaultState, { spaces: 2 });
    }
  }

  async readState(): Promise<DobbyState> {
    await this.ensureDataFile();
    const content = await fs.readJson(DATA_FILE);
    return {
      ...defaultState,
      ...content,
    } satisfies DobbyState;
  }

  async writeState(next: DobbyState): Promise<void> {
    await this.ensureDataFile();
    await fs.writeJson(DATA_FILE, next, { spaces: 2 });
  }

  async reset(): Promise<void> {
    await fs.remove(DATA_FILE);
  }
}

export const storage = new StorageService();

File: src/storage.rs
use std::fs;
use std::path::PathBuf;

use anyhow::{Context, Result};

use crate::state::{DobbyState, PlanRecord};

pub struct Storage {
    path: PathBuf,
}

impl Storage {
    pub fn new() -> Result<Self> {
        let mut dir = dirs::home_dir().context("Unable to locate home directory")?;
        dir.push(".dobby-cli");
        fs::create_dir_all(&dir).context("Failed to create Dobby state directory")?;
        Ok(Self {
            path: dir.join("state.json"),
        })
    }

    pub fn load(&self) -> Result<DobbyState> {
        if !self.path.exists() {
            return Ok(DobbyState::default());
        }
        let bytes = fs::read(&self.path).context("Failed to read Dobby state file")?;
        let state: DobbyState =
            serde_json::from_slice(&bytes).context("Failed to parse Dobby state file")?;
        Ok(state)
    }

    pub fn save(&self, state: &DobbyState) -> Result<()> {
        let contents =
            serde_json::to_vec_pretty(state).context("Failed to serialize Dobby state")?;
        fs::write(&self.path, contents).context("Failed to write Dobby state file")?;
        Ok(())
    }

    pub fn write_plan(&self, plan: PlanRecord) -> Result<()> {
        let mut state = self.load()?;
        state.plan = Some(plan);
        self.save(&state)
    }

    pub fn reset(&self) -> Result<bool> {
        if self.path.exists() {
            fs::remove_file(&self.path).context("Failed to remove state file")?;
            return Ok(true);
        }
        Ok(false)
    }
}
